azure/services/privatedns/privatedns.go

```diff
@@ -19,205 +19,252 @@ package privatedns
 
-	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
-	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/pkg/errors"
 
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
+const serviceName = "privatedns"
+
 // Scope defines the scope interface for a private dns service.
 type Scope interface {
 	azure.ClusterDescriber
-	PrivateDNSSpec() *azure.PrivateDNSSpec
+	azure.Authorizer
+	azure.AsyncStatusUpdater
+	PrivateDNSSpec() (zoneSpec azure.ResourceSpecGetter, linksSpec, recordsSpec []azure.ResourceSpecGetter)
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
-	Scope Scope
-	client
+	Scope              Scope
+	zoneClient         async.Creator
+	vnetLinkClient     async.Creator
+	zoneReconciler     async.Reconciler
+	vnetLinkReconciler async.Reconciler
+	recordReconciler   async.Reconciler
 }
 
 // New creates a new private dns service.
 func New(scope Scope) *Service {
+	zoneClient := newPrivateZonesClient(scope)
+	vnetLinkClient := newVirtualNetworkLinksClient(scope)
+	recordSetsClient := newRecordSetsClient(scope)
 	return &Service{
-		Scope:  scope,
-		client: newClient(scope),
+		Scope:              scope,
+		zoneClient:         zoneClient,
+		vnetLinkClient:     vnetLinkClient,
+		zoneReconciler:     async.New(scope, zoneClient, zoneClient),
+		vnetLinkReconciler: async.New(scope, vnetLinkClient, vnetLinkClient),
+		recordReconciler:   async.New(scope, recordSetsClient, recordSetsClient),
 	}
 }
 
 // Reconcile creates or updates the private zone, links it to the vnet, and creates DNS records.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.Reconcile")
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.Reconcile")
 	defer done()
 
-	zoneSpec := s.Scope.PrivateDNSSpec()
-	if zoneSpec != nil {
-		// Skip the reconciliation of private DNS zone which is not managed by capz.
-		isManaged, err := s.isPrivateDNSManaged(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "could not get private DNS zone state of %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-		}
-		// If resource is not found, it means it should be created and hence setting isVnetLinkManaged to true
-		// will allow the reconciliation to continue
-		if err != nil && azure.ResourceNotFound(err) {
-			isManaged = true
-		}
-		if !isManaged {
-			log.V(1).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zoneSpec.ZoneName)
-			log.V(1).Info("Tag the DNS manually from azure to manage it with capz."+
-				"Please see https://capz.sigs.k8s.io/topics/custom-dns.html#manage-dns-via-capz-tool", "private DNS", zoneSpec.ZoneName)
-			return nil
-		}
```

Comment:
I have moved this logic to `spec.Parameters` since we already do a GET to fetch the existing resource. Same for vnet links.